source/extensions/common/wasm/context.cc

```diff
@@ -0,0 +1,212 @@
+  root_local_info_ = &std::static_pointer_cast<Plugin>(plugin)->local_info_;
+}
+
+uint64_t Context::getCurrentTimeNanoseconds() {
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(
+             wasm()->time_source_.systemTime().time_since_epoch())
+      .count();
+}
+
+WasmResult Context::log(uint64_t level, absl::string_view message) {
+  switch (static_cast<proxy_wasm::LogLevel>(level)) {
+  case proxy_wasm::LogLevel::trace:
+    ENVOY_LOG(trace, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::debug:
+    ENVOY_LOG(debug, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::info:
+    ENVOY_LOG(info, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::warn:
+    ENVOY_LOG(warn, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::error:
+    ENVOY_LOG(error, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::critical:
+    ENVOY_LOG(critical, "wasm log{}: {}", log_prefix(), message);
+    break;
+  default:
+    return WasmResult::BadArgument;
+  }
+  return WasmResult::Ok;
+}
+
+WasmResult Context::getProperty(absl::string_view path, std::string* result) {
+  if (path.size() == 15 && !memcmp(reinterpret_cast<const void*>(&*path.begin()),
+                                   reinterpret_cast<const void*>("plugin_root_id\0"), 15)) {
+    std::cerr << "returning " << root_id() << "\n";
+    *result = std::string(root_id());
+    return WasmResult::Ok;
+  }
+  return WasmResult::Unimplemented;
+}
+
+WasmResult Context::defineMetric(MetricType type, absl::string_view name, uint32_t* metric_id_ptr) {
+  auto stat_name = wasm()->stat_name_set_->add(name);
+  if (type == MetricType::Counter) {
+    auto id = wasm_->nextCounterMetricId();
+    auto c = &wasm()->scope_->counterFromStatName(stat_name);
+    wasm()->counters_.emplace(id, c);
+    *metric_id_ptr = id;
+    return WasmResult::Ok;
+  } else if (type == MetricType::Gauge) {
+    auto id = wasm()->nextGaugeMetricId();
+    auto g = &wasm()->scope_->gaugeFromStatName(stat_name, Stats::Gauge::ImportMode::Accumulate);
+    wasm()->gauges_.emplace(id, g);
+    *metric_id_ptr = id;
+    return WasmResult::Ok;
+  } else if (type == MetricType::Histogram) {
+    auto id = wasm()->nextHistogramMetricId();
+    auto h = &wasm()->scope_->histogramFromStatName(stat_name, Stats::Histogram::Unit::Unspecified);
+    wasm()->histograms_.emplace(id, h);
+    *metric_id_ptr = id;
+    return WasmResult::Ok;
+  }
+  return WasmResult::BadArgument;
+}
+
+WasmResult Context::incrementMetric(uint32_t metric_id, int64_t offset) {
+  auto type = static_cast<MetricType>(metric_id & Wasm::kMetricTypeMask);
+  if (type == MetricType::Counter) {
+    auto it = wasm()->counters_.find(metric_id);
+    if (it != wasm()->counters_.end()) {
+      if (offset > 0) {
+        it->second->add(offset);
+        return WasmResult::Ok;
```

Comment:
The coverage build is broken right now, but please make sure all of the C-style error handling this file has full test coverage. I will be checking before merge. Thank you!

Reply:
I'll make a coverage pass once we get the first set of reviews done and have resolved any high level changes which might invalidate tests for low level details like error handling conditionals.